sklearn/pipeline.py

```diff
@@ -542,6 +542,108 @@ def _name_estimators(estimators):
+    A step's estimator may be replaced entirely by setting the parameter
+    with its name to another estimator, or a transformer removed by setting
+    to None.
+
+    Read more in the :ref:`User Guide <pipeline>`.
+
+    Parameters
+    ----------
+    steps : list
+        List of (name, transform) tuples (implementing fit/transform) that are
+        chained, in the order in which they are chained, with the last object
+        an estimator.
+
+    memory : Instance of joblib.Memory or string (optional)
+        Used to cache the output of the computation of the tree.
+        By default, no caching is done. If a string is given, it is the
+        path to the caching directory.
+
+    Attributes
+    ----------
+    named_steps : dict
+        Read-only attribute to access any step parameter by user given name.
+        Keys are step names and values are steps parameters.
+
+    Examples
+    --------
+    """
+
+    def __init__(self, steps, memory=None):
+        self.memory = memory
+
+        super(CachedPipeline, self).__init__(steps)
+
+    def _fit(self, X, y=None, **fit_params):
+        self._validate_steps()
+        fit_params_steps = dict((name, {}) for name, step in self.steps
+                                if step is not None)
+        for pname, pval in six.iteritems(fit_params):
+            step, param = pname.split('__', 1)
+            fit_params_steps[step][param] = pval
+        Xt = X
+        for name, transform in self.steps[:-1]:
+            if transform is None:
+                pass
+            else:
+                Xt, transform = self._fit_transform_one(
+                    transform, name,
+                    None, Xt, y,
+                    **fit_params_steps[name])
+        if self._final_estimator is None:
+            return Xt, {}
+        return Xt, fit_params_steps[self.steps[-1][0]]
+
+    def fit(self, X, y=None, **fit_params):
+        """Fit the model
+
+        Fit all the transforms one after the other and transform the
+        data, then fit the transformed data using the final estimator.
+
+        Parameters
+        ----------
+        X : iterable
+            Training data. Must fulfill input requirements of first step of the
+            pipeline.
+
+        y : iterable, default=None
+            Training targets. Must fulfill label requirements for all steps of
+            the pipeline.
+
+        **fit_params : dict of string -> object
+            Parameters passed to the ``fit`` method of each step, where
+            each parameter name is prefixed such that parameter ``p`` for step
+            ``s`` has key ``s__p``.
+
+        Returns
+        -------
+        self : Pipeline
+            This estimator
+        """
+        memory = self.memory
+        if isinstance(memory, six.string_types) or memory is None:
+            memory = Memory(cachedir=memory, verbose=10)
+        self._fit_transform_one = memory.cache(_fit_transform_one)
```

Comment:
Don't do that: don't decorate methods. Use the pattern described in https://github.com/joblib/joblib/blob/master/doc/memory.rst, under the bullet point "caching methods".